### ISSUE TITLE
Epic: GOLD knowledge distillation on BlueVela via SkyPilot (multi-node LSF)

### ISSUE THREAD

**Issue by cmadam** [cmadam/granite.build#61]
## Overview

Knowledge distillation of a small **student** from a larger **teacher** using the `kd-sandbox` GOLD
trainer (generalized JSD) is validated on BlueVela today — but only through `gbansible`. An ansible
role renders a `gold_config.yaml`, `scp`s it to the cluster, and runs `launch_enroot.sh` over SSH,
which `bsub`s and fans out with `blaunch`. There is no build record, no artifact, no recipe, and none
of the environment/step machinery is reused. Reference: `granite-dot-build/gbansible/docs/distillation.md`.

This epic makes it a first-class `gold-distill` step on the `skypilot/lsf/ibm-bluevela` environment.

The blocker is that **multi-node does not work through gbserver's SkyPilot path at all**:
`sky.Task` is constructed without `num_nodes` (`skypilot.py:1678`), and the pinned SkyPilot fork's
LSF cloud has multi-node *allocation* (`#BSUB -n N`, `span[ptile=1]`, `blaunch`, topology export) but
a single-node *execution* layer (`LsfCodeGen` hardcodes `SKYPILOT_NUM_NODES=1`, `LsfCommandRunner`
has no node targeting, the dispatch dir is one shared path every node `rm -rf`s, and there is no
`sky/skylet/executor/lsf.py`). Distillation needs >= 2 nodes — 4 in the reference run — so both
layers are in scope.

## Issues

- [ ] #57 — SkyPilot fork: generic multi-node support in the LSF cloud
- [ ] #58 — gbserver: thread `num_nodes` into `sky.Task` for the SkyPilot launcher
- [ ] #59 — New `gold-distill` step for skypilot/lsf
- [ ] #60 — BlueVela GOLD recipes: 2-node smoke + off-policy reference

## Decisions

1. **BlueVela LSF via SkyPilot only.** The Vela/K8s PyTorchJob path (in `karve/assets.git`) is out of
   scope; the BlueVela enroot route is the validated one and the better perf choice for large
   multinode runs.
2. **Deps from the image, trainer from `/proj`.** The container image
   (`kd-sandbox-distill:0.1.0-uv`, venv at `/stage/.venv`) carries the pinned stack; the trainer comes
   from the live `/proj/granite-build/g4os/kd-sandbox` checkout, which the provisioner already
   identity-mounts. This is what the validated run already does, and a trainer fix then needs no
   granite.build change. Cost: the checkout is unpinned mutable state, so the step records
   `git -C $KD_CODE_DIR rev-parse HEAD`.
3. **Full generic multi-node fix in the fork**, not just threading `num_nodes`. The minimal path
   (letting the step consume the fork's already-exported `RANK`/`MASTER_ADDR`) would work for this
   one workload, but leaves `SKYPILOT_NODE_RANK` broken for every other.
4. **Checkpoints under `$GB_BUILD_WORKDIR`**, published as a `model` artifact — not the launcher's
   hardcoded `$SHARE_PATH/kd-sandbox/checkpoints/`.
5. **First milestone: off-policy, 2 nodes** (`lmbda=0`, no vLLM), with `grad_accum=12` holding
   effective batch at 192.
6. **Smoke with a small teacher/student and tiny dataset**, minutes not hours. Smoke and reference
   differ only in `parameters.yaml` values.
7. **On-policy is a later phase.** It continues from a good off-policy checkpoint, so it cannot even
   be configured until off-policy has produced one. The design must not preclude it — and does not:
   the vLLM/trainer split becomes a step-level branch on `RANK` with zero further fork changes.

## Delivery model

Epic branch = integration branch, fresh from `upstream/main`
(`gbserver-worktrees/epic-gold-kd-bluevela-skypilot`). Each milestone branches from the epic, PRs on
`origin`, merges back into the epic; one PR against `upstream/main` at the end.

The #57 work lives in `cmadam/skypilot` on `feature/lsf-multinode`, branched from and merged into
`granite-build` — the branch `pyproject.toml` pins. Until that merge, gbserver develops against the
fork with `pip install -e /home/cma/de/cma/skypilot` plus `sky api stop` after every fork edit (the
API server is a separate long-lived process that imports `sky` once).

Landing order: A1-A4 -> #58 -> A5 -> **live 2-node smoke test** -> A6-A8 -> #59 -> **live GOLD smoke**
-> A9 (merge + new tag `gb-sky-v2-multinode`) -> repin `pyproject.toml` -> #60 -> **4-node then
2-node reference**.

## Verification

- **#57:** golden-file diffs per commit; the two `single_node_*` goldens must not move. SLURM tests
  run every commit (A7/A8 touch shared `command_runner.py`, `provisioner.py`, `instance_setup.py`).
- **#58:** `GBTEST_MODE=mock` — precedence is pure config computation, no cluster needed.
- **The single highest-value live test**, after A5 and #58: a 2-node `command` step echoing
  `$(hostname) rank=$RANK/$TOTAL_NODES`. Success = **two distinct hostnames and both ranks**, for a
  few minutes of queue time. If it prints `0,0`, stop — that is the rank-derivation hazard below.
- **#59:** `make space` / `make test` / `make check-published`, and the config-renderer unit tests
  (the `%.2e` learning-rate rule is the cheapest place to catch the most expensive bug).
- **#60:** live BlueVela via gbmcp (`build_start` -> `build_status` -> `build_job_log`), 4-node before
  2-node.

## Principal risks

- **ZeRO-3 memory ceiling at 2 nodes.** 16 ranks instead of 32 doubles the per-GPU shard, and
  `per_device_train_batch_size` is already 1 with gradient checkpointing on — no headroom lever left
  but `max_length`. Detected only at the 2-node reference run, which is why 4-node runs first.
- **Rank/host disagreement.** `get_cluster_info`'s rank tag (from `bjobs -o EXEC_HOST` order) and the
  in-job `RANK` (from `$LSB_HOSTS` order) are uncorrelated. A wrong `MASTER_ADDR`/rank pairing hangs
  NCCL for the full `TORCH_NCCL_TIMEOUT_MS=3600000` — one silent hour per attempt. Designed out by
  keying on hostname and having the job publish a rank manifest.
- **Silent NCCL TCP fallback** — a run that "works" at a fraction of throughput, with nothing failing.
- **Fork/gbserver version skew** — an older sky silently ignores `num_nodes` and "succeeds" with a
  garbage checkpoint. Converted to a startup error by #58's preflight guard.
- **`~/.lsf/config` refuse-on-conflict** — editing an environment's `cluster_ssh_configs` hard-fails
  every build until the file is hand-deleted. Avoided entirely: nothing here touches it.

**Comment by cmadam** [comment]
## Finding: granite 4.x `tokenizer_class` mis-segments under `AutoTokenizer`, and every recorded eval row is affected

Measured while retargeting the distillation pipeline from the 4.1-3b/4.2-30b reference pair to a granite-4.0-350m student with a granite-4.1-3b teacher. Two cheap probe builds on BlueVela, no training.

### What

Every granite 4.x checkpoint declares `tokenizer_class: "GPT2Tokenizer"` in `tokenizer_config.json`, while its `tokenizer.json` carries

```
pre_tokenizer = Sequence[ Split(Regex "(?i:'s|'t|...|\p{N}{1,3}|..."), ByteLevel(use_regex=False) ]
```

`AutoTokenizer` honours the declared class and substitutes `ByteLevel(use_regex=True)` — GPT-2's own regex — discarding the trained `Split`. This is the trap `distill-tokenizer-align` was written for (M3, #59); what is new is that it is **not** confined to the student, and that we now have a measurement of which side the weights prefer.

Build `b70cd903`, on the granite-4.0-350m SFT checkpoint:

```
PROBE declared_class GPT2Tokenizer
PROBE as-is  class=GPT2Tokenizer     pre_tokenizer=ByteLevel(use_regex=True)
PROBE pinned class=TokenizersBackend pre_tokenizer=Sequence[Split(regex), ByteLevel(use_regex=False)]
PROBE DIFF "don't 1234 tokenize"
      as-is  [15357, 956, 220,  717, 1958, 78751]     -> 12|34
      pinned [15357, 956, 220, 4513,   19, 78751]     -> 123|4
```

The clearest case is digit grouping: the trained regex splits digit runs at `\p{N}{1,3}`, GPT-2's `\p{N}+` groups them. But it is broader than digits — over 200 real conversations the two produce **272,276 vs 264,684 tokens for the same 1,326,879 bytes** (2.8% apart), which digit runs alone cannot account for in conversational text.

### Which segmentation the weights actually want

Build `8803bcf9`, same checkpoint, 200 conversations, scored under both tokenizers:

| | class | tokens | NLL/byte |
|---|---|---|---|
| as declared | `GPT2Tokenizer` | 272,276 | 0.58593 |
| `tokenizer_class` pinned | `TokenizersBackend` | 264,684 | **0.52224** |

**The model prefers the pinned tokenizer by 10.9% NLL/byte** (PPL/byte 1.797 → 1.686), and the pinned one is 2.8% more token-efficient.

Compared **per byte, not per token**: two segmentations yield different token counts, so perplexity-per-token favours the coarser one by construction. Worth flagging that the "26.1 vs 3.29 PPL/token" figure in `run-align.sh`'s header has that normalisation problem — a gap that size likely survives it, but the units are not comparable as stated.

The direction is consistent with pretraining having used `tokenizer.json`: the embeddings and BPE merges are coherent under the trained pre_tokenizer, and the SFT stage — which trained on a pre-tokenized dataset, most likely produced via `AutoTokenizer` — introduced a mismatch. **Pinning is a fix, not a regression.**

### Why it matters for this epic

`distill-hf-export` pins `tokenizer_class` (see `adfdc64`). The recorded baseline eval rows do not — the sage harness loads with `AutoTokenizer`. So comparing a distilled model against a recorded row compares **two tokenizers as well as two models**, and would credit distillation with a tokenizer fix it did not perform.

Consequences:

1. `distill-hf-export`'s pin is correct; keep it, no override needed.
2. Any distillation-vs-baseline comparison needs a **tokenizer-control column**: `full-eval` on the *same* weights with only `tokenizer_class` rewritten (27 preemptable 1×H100 jobs, no training). It is a standalone result — what pinning alone recovers at zero training cost.
3. Likely wider than this epic: every recorded granite4-350m row (after-SFT, after-IFRL, after-IDRL) was produced through `AutoTokenizer`, so all of them are probably **understated**. Same applies to whatever produced `datasets/tokenized/all_combined/...`, i.e. the SFT training data itself.

Note `tokenizer_class: PreTrainedTokenizerFast` resolves to a runtime class named `TokenizersBackend` under transformers 5.8.0. That is the value `adfdc64` exists to rewrite on export, since transformers 4 (the BFCL image) cannot resolve it.

Also measured, same probe (build `b70cd903`): **the `granitemoehybrid` / mamba-ssm load failure from build `96868683` does not reproduce** — `granite-4.0-350m` loads under both `flash_attention_2` and `eager` in `kd-sandbox-distill:0.1.0-uv` (352,379,904 params). No `HF_HOME` prefetch needed. Worth noting `logits_scaling` is 4 on the 350m student against 10.0 on the 3b teacher, so the effective KD temperature is not the `1.0` such a config states.

### Also fixed

`run-align.sh` passed `--require-chatml` for the teacher unconditionally, so a granite-4.1 teacher aborted in stage `[1/4]` — no granite 4.x model has `<|im_start|>` in `added_tokens_decoder`. Added `align_config.require_chatml` (default `true`, reference pair unchanged). It needed no upstream Python change: `build_overlay` already accepted both flag forms and `verify()` already took the keyword, so `code_config.expect_ref` is untouched.

Reproduce any of the above with `recipes/granite4-350m/lsf/distill-probe/`, whose README carries these numbers and the verdict tables.

**Comment by cmadam** [comment]
## Stage 1 run verification — build `df8512e0-be93-4762-85ad-813d9080d72f`

`recipes/granite4-350m/lsf/distill-stage1/build.yaml`, launched as:

```
gb build start -f recipes/granite4-350m/lsf/distill-stage1/build.yaml --space public \
  --param NCCL_DEBUG=INFO --param GOLD_SAVE_STEPS=250
```

All 8 targets reported SUCCESS. I audited the seven pre-BFCL targets against their real
workload stdout rather than trusting the target status. **Every one of them ran correctly.**
Both `train-gold` ranks exited 0, as did every other step.

Note for anyone repeating this: the real workload stdout is in
`/proj/granite-build/g4os/skypilot/gb-<hash>-<suffix>/.sky/dispatch/<host>/out_run-*.log`,
with the exit code in the sibling `rc_run-*`. `~/sky_logs` does not exist on the login node
and `build_job_log`'s workdir glob finds nothing for these builds.

### Per-target findings

**sources** — Proportional sampling worked as designed: 15.7434% of each split
(general 678,849 / tools 111,824 / rag 9,327 = 800,000 rows) out of 5,081,504 total.
`renamed=5081504 already_messages=0 bad=0` — every row carried `conversations` and was
renamed; nothing dropped as unparseable.

**align** — All 4 stages verified, 6/6 round-trip probes, pre_tokenizer preserved
(`Sequence`). The retag was vacuous exactly as the recipe predicted:
`reuse_in_place=100352 moved=0 transplant=0 mean_init=0`. The masking contract agrees on
all 6 template probes (1/1/1/1/2/2 spans, boundary delta uniformly `[-1]`), under both
`thinking=True` and `thinking=False`. `require_chatml: false` behaved as intended — the
step recorded that the teacher has no `<|im_start|>` but has a vocabulary identical to the
student's, so the role map is vacuous rather than inapplicable.

**corpus** — Internally consistent: 800,000 in → 13,680 dropped
(7,772 `last_message_not_assistant` + 5,908 `over_max_length`) → 786,320 kept →
782,388 train + 3,932 eval. `eval_fraction: 0.005` lands exactly.

**train-gold** — Exactly 8,150 optimizer steps = one full epoch:
782,388 ÷ (2 × 3 × 16 GPUs = 96) = 8,149.9. Off-policy is genuinely off-policy:
`t_gen=0`, `n_sync_params=0`, `use_vllm: False`, and `off_policy_loss == loss` on every
one of the 8,150 logged steps. Loss 0.0825 → `train_loss 0.06879`; grad_norm 0.86 → 0.53;
LR warms over 5% to 1.989e-06 then cosines to the 2e-07 floor. No NaN, no OOM, no NCCL
fault. 13h 20m wall clock on 2 nodes × 8 H100.

**export** — Selected `checkpoint-8150`, kept 6 files, dropped the 16
`rng_state_<rank>.pth` files exactly as the recipe's `allow_unknown: true` comment
anticipated.

**eval-transfer-baseline / eval-transfer** — Genuinely apples-to-apples: same 256 rows,
same 102,056 assistant tokens, `truncated_at_max_length: 0`,
`skipped_no_assistant_span: 0` on both sides.

| metric | baseline (t=0) | post-training | Δ |
|---|---|---|---|
| JSD | 0.08283 | 0.06802 | **−17.9%** |
| KLD | 1.03062 | 0.65639 | **−36.3%** |
| RKLD | 0.43836 | 0.43062 | −1.8% |
| entropy | 0.68357 | 0.39562 | −42.1% |

### Three things worth attention

**1. The alarming line in the LSF log is not an error.** `train-gold`'s
`sky_logs/1874192.out` ends with `TERM_OWNER: job killed by owner. Exited with signal
termination: 9`. That is SkyPilot bkilling the allocation after the workload finished.
Both `rc_run-*` files are `0`. Check `rc_run-*`, not the LSF termination reason.

**2. The exported `tokenizer.json` carries stale trainer runtime state.** This is why the
post-training eval reports `agreement: ids_agree_despite_differing_tokenizer_json` while
the baseline reports `identical_tokenizer_json`. Vocab, merges, pre_tokenizer and
added_tokens are all identical; two keys differ from align's:

- `truncation`: `null` → `{"direction":"Right","max_length":466,"strategy":"LongestFirst","stride":0}`
  — a 466-token truncation budget baked into the shipped model. transformers'
  `PreTrainedTokenizerFast` sets truncation per call, so nothing in this run was affected
  (both evals report `truncated: 0`), but a consumer loading it through the raw
  `tokenizers.Tokenizer.from_file()` backend inherits that limit silently.
- `post_processor`: `null` → an identity `TemplateProcessing` (`single=[A]`, `pair=[A,B]`,
  `special_tokens={}`). Adds nothing; harmless.

Export's `verify` passes because it compares vocab/ids, not bytes. `distill-hf-export`
already normalises `tokenizer_config.json` (tokenizer_class, padding_side,
local_files_only, is_local); clearing `truncation` and `padding` in `tokenizer.json` too
would be a one-line addition to the same normalisation list.

**3. RKLD is flat (−1.8%) while entropy fell 42%.** Forward KL improved 36% but reverse KL
did not move. The student sharpened rather than broadened — concentrating mass where it
already agreed with the teacher instead of covering the teacher's spread. That is the
expected shape for β=0.5 JSD at this LR, but it is the number to watch in stage 2: if
entropy keeps collapsing while RKLD stalls, the student is narrowing rather than learning
the teacher's distribution. It is also the plausible mechanism behind the degenerate
repetition loops seen in the BFCL generations.

### Two smaller notes

- The corpus step's own `NOTE` flags that `completion_boundary=last_message` leaves
  34,650,329 assistant tokens (10.2%) unsupervised. Deliberate, and coupled to the eval
  span, but a real cost worth revisiting for stage 2.
- The build is still named `distill-350m-smoke`, and the build.yaml header describes a
  smoke recipe, while this is the stage-1 full-epoch run. Cosmetic, but it makes
  `gb build list` confusing.
